tool/tsh/common/tsh.go

```diff
@@ -2086,14 +2086,36 @@ func onLogout(cf *CLIConf) error {
 		proxyHost = cf.Proxy
 	}
 
-	switch {
-	// Proxy and username for key to remove.
-	case proxyHost != "" && cf.Username != "":
-		tc, err := makeClient(cf)
+	tc, err := makeClient(cf)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	var SAMLSingleLogoutURL string
+	clt, err := tc.ConnectToCluster(cf.Context)
+	if err != nil {
+		log.
+			Warnf("Failed to retrieve cluster client for user %s, SAML single logout will be skipped: %v", active.Username, err)
+	}
+
+	var user types.User
+	// Only run this if the ConnectToCluster above was successful.
+	if err == nil {
+		user, err = clt.AuthClient.GetUser(cf.Context, active.Username, false)
 		if err != nil {
-			return trace.Wrap(err)
+			log.
+				Warnf("Failed to retrieve user details for user %s, SAML single logout will be skipped: %v", active.Username, err)
 		}
+	}
+	if user != nil {
+		if len(user.GetSAMLIdentities()) > 0 {
+			SAMLSingleLogoutURL = user.GetSAMLIdentities()[0].SAMLSingleLogoutURL
```

Comment:
A utility function that acts on a `TeleportClient` would make this whole logic far more easy to check and follow, as you'd be able to early return on errors.